File: notify-on-duty/notify-on-duty.pl
#!/usr/bin/perl
# nagios: +epn
#
# notify-on-duty.pl - Notify script with own MySQL DB to notify
#                     over SMS the right contact to a certain time.
#
# Copyright (C) 2018 Juergen Vigna
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
#
# Report bugs to:  juergen.vigna@wuerth-phoenix.com
#
#
use strict;
use warnings;
use vars qw($PROGNAME);

use Data::Dumper;
use Getopt::Long;
use Date::Parse;
use DBI;
use POSIX;
use File::Basename;

$PROGNAME = basename($0);
my $VERSION  = "1.0.0";
sub send_notification;
sub print_help;
sub print_usage;

my ($opt_h,$opt_V);
my $team = undef;
my $defaultpager = undef;
my $message = undef;
my $datetime = undef;
my $mysql_host = "localhost";
my $mysql_onduty_database  = "notify_onduty";
my $mysql_monarch_database = "monarch";
my $mysql_username = "notify";
my $mysql_password = "onduty";
my $opt_verbose = undef;
my @opt_debug;
my $opt_testonly = undef;
my $notifycmd = "/etc/nagios/neteye/sms/smssend.sh '\@CONTACT\@' '\@MESSAGE\@'";

use Getopt::Long;
&Getopt::Long::config('bundling');
GetOptions(
        "V"   => \$opt_V,        "version"          => \$opt_V,
        "h"   => \$opt_h,        "help"             => \$opt_h,
	"T=s" => \$team,         "team=s"           => \$team,
	"P=s" => \$defaultpager, "pager=s"          => \$defaultpager,
	"M=s" => \$message,      "message=s"        => \$message,
	"d=s" => \$datetime,     "datetime=s"       => \$datetime,
	"N=s" => \$notifycmd,    "notifycmd=s"      => \$notifycmd,
                                 "mysql_host=s"     => \$mysql_host,
                                 "mysql_database=s" => \$mysql_onduty_database,
                                 "mysql_username=s" => \$mysql_username,
                                 "mysql_password=s" => \$mysql_password,
        "v"   => \$opt_verbose,  "verbose"          => \$opt_verbose,
        "D"   => \@opt_debug,    "debug"            => \@opt_debug,
        "test" => \$opt_testonly, help='Test mode, it only shows the message but does not send it'  
);

# -h & --help print help
if ($opt_h) { print_help(); exit 0; }
# -V & --version print version
if ($opt_V) { printf "%s, Version %s\n",$PROGNAME, $VERSION; exit 0; }

if (!defined($team)) {
	print_help();
	print "Please specify the team to send this notification to, aborting!\n";
	exit 1;
}

if (!defined($message)) {
	print_help();
	print "Please specify the message to send in this notification, aborting!\n";
	exit 1;
}

my $dbnotify  = "DBI:mysql:database=$mysql_onduty_database;host=$mysql_host";
my $dbmonarch = "DBI:mysql:database=$mysql_monarch_database;host=$mysql_host";
my $dbn = DBI->connect($dbnotify, $mysql_username, $mysql_password);
my $dbm = DBI->connect($dbmonarch, $mysql_username, $mysql_password);

my $qn;
my $qni;
my $qm;
my $ref;

if (defined($datetime)) {
	$qn = $dbn->prepare("SELECT datetime FROM schedules WHERE team = '$team' AND datetime <= '$datetime' ORDER BY datetime DESC LIMIT 1")
	  or die "Cannot prepare mysql query for DB ($mysql_onduty_database): $dbn->errstr()";
} else {
	$qn = $dbn->prepare("SELECT datetime FROM schedules WHERE team = '$team' AND datetime <= NOW() ORDER BY datetime DESC LIMIT 1")
	  or die "Cannot prepare mysql query for DB ($mysql_onduty_database): $dbn->errstr()";
}

$qn->execute()
        or die "Cannot query mysql in DB ($mysql_onduty_database): $dbn->errstr()";

if ($qn->rows < 1) {
	if (defined($defaultpager)) {
		send_notification($team, undef, $defaultpager, $notifycmd, $message);
		if ($opt_verbose) {
			print "WARNING: Cannot find any entry in schedules table, notification sent to defaultpager ($defaultpager)!\n";
		}
		exit 0;
	}
	if ($opt_verbose) {
		print "ERROR: Cannot find any entry in schedules table, no notification sent!\n";
	}
	exit 1;
}

($datetime) = $qn->fetchrow_array();

#
# Now search all entries with this datetime and this team
#

$qn = $dbn->prepare("SELECT * FROM schedules WHERE team = '$team' AND datetime = '$datetime' ORDER BY contact_id")
  or die "Cannot prepare mysql query for DB ($mysql_onduty_database): $dbn->errstr()";
$qn->execute()
  or die "Cannot query mysql in DB ($mysql_onduty_database): $dbn->errstr()";

my $contact_id;
my $pager;
while ($ref = $qn->fetchrow_hashref()) {
        if (@opt_debug) {
                printf "%s\n", Data::Dumper::Dumper($ref);
        }
        $contact_id = $ref->{'contact_id'};
	send_notification($team, $contact_id, $defaultpager, $notifycmd, $message);
}

exit 0;
#
# FUNKTIONS
#
sub send_notification() {
	my ($team, $contact_id, $defaultpager, $sendcmd, $message) = @_;
	my $prefix = "";
	my $sending = 0;

	if (defined($contact_id)) {
		$qm = $dbm->prepare("SELECT pager FROM contacts WHERE contact_id = '$contact_id'")
		  or die "Cannot prepare mysql query for DB ($mysql_onduty_database): $dbm->errstr()";
		$qm->execute()
		  or die "Cannot query mysql in DB ($mysql_onduty_database): $dbm->errstr()";
		if ($qm->rows > 0) {
			($pager) = $qm->fetchrow_array();
			$sending = 1;
		} elsif (defined($defaultpager)) {
			$pager = $defaultpager;
			$prefix = "CONTACT_ID($contact_id) NOT FOUND: ";
			$contact_id = 9999;
			$sending = 1;
			if ($opt_verbose) {
				print "Notifying to Default Pager ($pager), contact_id ($contact_id) not found in DB\n";
			}
		}
	} elsif (defined($defaultpager)) {
		$pager = $defaultpager;
		$contact_id = 9999;
		$sending = 1;
		if ($opt_verbose) {
			print "Notifying to Default Pager ($pager)\n";
		}
	}
	if (!$sending) {
		if (defined($opt_testonly)) {
			print "NOT notifying anything, as contact_id could not be found and no defaultpager\n";
			return;
		}
		$qni = $dbn->prepare("INSERT INTO notify_log (timestamp, team, contact_id, message) VALUES (NOW(), '$team', '0', '$message')")
	          or die "Cannot prepare mysql insert for DB ($mysql_onduty_database): $dbn->errstr()";
		$qni->execute()
	          or die "Cannot insert into mysql in DB ($mysql_onduty_database): $dbn->errstr()";
		return;
	}
	if (defined($opt_testonly)) {
		print "Notifying Contact ($contact_id) on Pager ($pager)\n";
		return;
	}
	$sendcmd =~ s/\@CONTACT\@/$pager/g;
	$sendcmd =~ s/\@MESSAGE\@/$message/g;
	`$sendcmd`;
	$qni = $dbn->prepare("INSERT INTO notify_log (timestamp, team, contact_id, message) VALUES (NOW(), '$team', '$contact_id', '$prefix$message')")
          or die "Cannot prepare mysql insert for DB ($mysql_onduty_database): $dbn->errstr()";
	$qni->execute()
          or die "Cannot insert into mysql in DB ($mysql_onduty_database): $dbn->errstr()";
}

sub print_help() {
        printf "%s, Version %s\n",$PROGNAME, $VERSION;
        print "Copyright (c) 2018 Juergen Vigna <juergen.vigna\@wuerth-phoenix.com>\n";
        print "This program is licensed under the terms of the\n";
        print "GNU General Public License\n(check source code for details)\n";
        print "\n";
        printf "Script to send notifications over a schedule table in MySQL\n";
        print "\n";
        print_usage();
        print "\n";
        print " -V (--version)      Plugin version\n";
        print " -h (--help)         usage help\n";
        print " --test              Test Mode only, do not send anything\n";
		print " -T (--team)         Name of team to send this notification (required)\n";
		print " -P (--defaultpager) Default Pager to send notification to if team or contact_id could not be found\n";
		print " -M (--message)      Text of message to send for this notification (required)\n";
		print " -d (--datetime)     MySQL Datetime to send this notification for (default: NOW())\n";
		print " -N (--notifycmd)    The command to use for this notification\n";
		print "                     Use this 2 macros inside the command:\n";
		print "                     \@CONTACT\@: Is substituted by the actual contact(s)\n";
		print "                     \@MESSAGE\@: Is substituted by your message\n";
		print "                     default: $notifycmd\n";
        print "\n";
        print " MySQL Integration:\n";
        print " --mysql_onduty_database mysql database name (default $mysql_onduty_database)\n";
        print " --mysql_username        mysql database username (default $mysql_username)\n";
        print " --mysql_password        mysql database password (default ******)\n";
        print " --mysql_host            mysql database hostname/ip (default $mysql_host)\n";
        print "\n";
}

sub print_usage() {
        print "Usage: \n";
        print "  $PROGNAME -T|--team <teamname> -M <message> [-P <defaultpager>] [-d <datetime>] [-N <nofifycmd>] [--mysql_onduty_database <databasename>] [--mysql_username <username>] [--mysql_password <password>] [--mysql_host <host/ip>]\n";
        print "  $PROGNAME [-h | --help]\n";
        print "  $PROGNAME [-V | --version]\n";
}

File: notify-on-duty/notify-on-duty.py
#!/usr/bin/python3

# ./notify-on-duty.py - Notify script with own MySQL DB to notify
#                     over SMS the right contact to a certain time.
#
# Copyright (C) 2022 Marco Ettocarpi, WuerthPhoenix GmbH
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
#
# Report bugs to:  net.support@wuerth-phoenix.com
#

import datetime
import argparse
import sys
import mysql.connector 
import os
from contextlib import closing
from  constants import *

version = '2.0.0'
parser = argparse.ArgumentParser()
# parser.add_argument("ls")
parser.add_argument('-V', '--version', action='version', version=version)
# parser.add_argument('-h','--help')
parser.add_argument('-T', '--team')
parser.add_argument('-M', '--message')
parser.add_argument('-d', '--datetime')
parser.add_argument('-N', '--notifycmd')
parser.add_argument('-P', '--defaultpager')

parser.add_argument('-mysql_host')
parser.add_argument('-mysql_database')
parser.add_argument('-mysql_username')
parser.add_argument('-mysql_password')


parser.add_argument('-v', '--verbose')
parser.add_argument('-D', '--debug')
parser.add_argument('-test',action='store_true',help='Run in dry mode. Do not send out any message')

args = parser.parse_args()


team = args.team
message = args.message
comment = ""

# db connection
if args.mysql_host is not None:
    host = args.mysql_host
else:
    host = MYSQL_HOST
if args.mysql_database is not None:  
    db = args.mysql_database
else:
    db = MYSQL_DATABASE
if args.mysql_username is not None:    
    user = args.mysql_username
else:
     user = MYSQL_USERNAME    
if args.mysql_password is not None:   
    passwd = args.mysql_password
else:
    passwd = MYSQL_PASSWORD
    
if args.test:
    comment = "TEST mode ON"

defaultpager = args.defaultpager

NOTIFYCMD = "/usr/bin/smssend"
#/usr/bin/smssend.sh

def conversion(sec):
    sec_value = sec % (24 * 3600)
    hour_value = sec_value // 3600
    sec_value %= 3600
    min = sec_value // 60
    sec_value %= 60
    return hour_value,min
   
def eqdate(date1,date2):
    return date1.year==date2.year and date1.month==date2.month and date1.day == date2.day

def usage():
    print("Usage: ")
    print("  ./notify-on-duty.py -T <team> -M <MESSAGE>")
    print(" ")
    print("For help call: ")
    print("  ./notify-on-duty.py --help")

def main():
    # if(args.version):
    #         print("%s, Version %s\n",version)
    if(args.team is None):
        print("Please specify the team to send this notification to, aborting!\n")
        usage()
        sys.exit(1)
    if(args.message is None):
        print("Please specify the message to send in this notification, aborting!\n")
        usage()
        sys.exit(1)
    datetime_str = args.datetime

    try:
        schedules = None
        if(datetime_str is not None):

            if(args.verbose):
                print("SELECT * FROM schedule WHERE team_id = (SELECT id  FROM team WHERE NAME='{}' )AND start_date = '{}' order by start_time".format(team, datetime_str))

            schedules = execute_sql_objects(
                "SELECT * FROM schedule WHERE team_id = (SELECT id  FROM team WHERE NAME='{}' )AND start_date = '{}' order by start_time".format(team, datetime_str))
        else:
            if(args.verbose):
                print("SELECT * FROM schedule WHERE team_id = (SELECT id  FROM team WHERE NAME='{}' )AND DATE_FORMAT(start_date, '%Y-%m-%d') = DATE_FORMAT(now(),'%Y-%m-%d') order by start_time".format(team))
            schedules = execute_sql_objects("SELECT * FROM schedule WHERE team_id = (SELECT id  FROM team WHERE NAME='{}' )AND DATE_FORMAT(start_date, '%Y-%m-%d') = DATE_FORMAT(now(),'%Y-%m-%d') order by start_time".format(team))

        if(args.verbose):
            print(schedules)

        if(len(schedules) < 1):

            # No schedule found: verify if send out notification to default pager
            if(args.defaultpager is not None):
                # send notification 
                send_notification(team, None, args.pager, NOTIFYCMD, message)
                print("defaultpager is not None")
                if(args.verbose):
                    print("WARNING: Cannot find any entry in schedules table, notification sent to defaultpager ({}})!\n".format(
                        args.defaultpager))
                    return
            if(args.verbose):
                print(
                    "ERROR: Cannot find any entry in schedules table, no notification sent!\n")
                return
        
        actual_date = datetime.datetime.now()
        #actual_date = datetime.datetime(2021,11,26,8,1 ,1)
        condition = False
        items = dict()
        for count,schedule in enumerate(schedules):
            # Old Logic
            #condition =  eqdate(schedule['start_date'],actual_date) and (conversion(schedule['start_time'].seconds)[0] <= actual_date.hour  and conversion(schedule['start_time'].seconds)[1] <= actual_date.minute)
            # Condition:
            # - schedule has date of today
            # - hour start IS LESS than current hour
            # - hour start IS EQUAL than current hour AND minute start IS LESS than current minute
            condition =  eqdate(schedule['start_date'],actual_date) and ((conversion(schedule['start_time'].seconds)[0] < actual_date.hour) or (conversion(schedule['start_time'].seconds)[0] == actual_date.hour) and (conversion(schedule['start_time'].seconds)[1] <= actual_date.minute))
            if(args.verbose):
                print("Condition match: {} Schedule start hour 0: {}, Schedule start minute 1: {}".format(condition,conversion(schedule['start_time'].seconds)[0],conversion(schedule['start_time'].seconds)[1]))
            if condition:
                items[count] = schedule
                if(args.verbose):
                    print("Choose: {}".format(schedule))
                # Do not break, as we should iterate through all matches until the last one since ordered by time
                #break
                   
        
        #
        # Now search all entries with this datetime and this team
        #
        #schedules =execute_sql_objects("SELECT * FROM schedule WHERE  team_id = (SELECT id  FROM team WHERE NAME='{}' ) AND start_date = '{}' ORDER BY team_id".format(team,datetime))
        schedule = schedules[len(items)-1]
        
        contact_id = None
        pager = None
        # for schedule in schedules:
        #     #if(opt_debug):
        #     if(args.test):
        #         print("%s\n")
        #     send_notification(team, schedule, defaultpager, NOTIFYCMD, message)
        # print("*** END ***")
        send_notification(team, schedule, defaultpager, NOTIFYCMD, message)

        if(args.verbose):
            print("*** END ***")
    except Exception as err:
        print(err)


def send_notification(team, schedule, defaultpager, sendcmd, message):
    
    if(args.verbose):
        print("SENDING NOTIFICATION")

    prefix = ""
    sending = False
    pager = ''
    # if user_phone_number not none => sending = True
    if schedule['user_phone_number'] is not None:
        sending = True
        pager = schedule['user_phone_number']
    elif defaultpager is not None:
        pager = defaultpager
        user = '9999'
        sending = True
        prefix  = "Phone number NOT FOUND:"
        if args.verbose:
            print("Notifying to Default Pager {}".format(pager)) 

    # If no user phone number identified
    if(not sending):
        print("ERROR: not able to identify phone number for contact ({} (ID:{})) on Pager ({})\n".format(schedule['user_name'],schedule['user_id'],pager))
        return

    # If user phone number identified
    if(sending):

        try:
           execute_sql_insert_upd("INSERT INTO notify_log (timestamp, team, contact_id, contact_name, phone_number, message, comment) VALUES (NOW(), '{}', '{}','{}','{}','{}','{}')".format(team,schedule['user_id'],schedule['user_name'],pager,message,comment))
        except Exception as e:
           print(e)      

        if args.test:
           print("Dry test mode: Would notify Team ID: {} Contact: {} (ID:{})) on Number: ({})\n".format(schedule['team_id'],schedule['user_name'],schedule['user_id'],pager))
           return
        #really execute send command
        else:
            sendcmd = NOTIFYCMD + " " + pager + " " + message
            os.system(sendcmd)

        return

    
    
def execute_sql(sql,json_str=False,test=False):
    """[summary]

    Args:
        sql ([string]): [sql query]

    Returns:
        [array]: [list of objects from db]
    """
    if test == True:
        return sql;
    try:
        with closing(mysql.connector.connect(host=host, db=db, user=user, passwd=passwd)) as my_conn:
            with closing(my_conn.cursor()) as my_curs:
                my_curs.execute(sql)
                return my_curs.fetchall()
    except mysql.connector.Error as err:
        print(err)   

def execute_sql_insert_upd(sql,json_str=False,test=False):
    """[summary]

    Args:
        sql ([string]): [sql query]

    Returns:
        [array]: [list of objects from db]
    """
    if test == True:
        return sql;
    try:
        with closing(mysql.connector.connect(host=host, db=db, user=user, passwd=passwd)) as my_conn:
            with closing(my_conn.cursor()) as my_curs:
                my_curs.execute(sql)
                my_conn.commit()
                return my_curs.lastrowid 
    except mysql.connector.Error as err:
        raise err

def execute_sql_objects(sql):
    with closing(mysql.connector.connect(host=host, db=db, user=user, passwd=passwd)) as my_conn:
            with closing(my_conn.cursor()) as my_curs:
                my_curs.execute(sql)
                rows = [x for x in my_curs]
                cols = [x[0] for x in my_curs.description]
                objects = []
                for row in rows:
                    object = {}
                    for prop, val in zip(cols, row):
                        object[prop] = val
                    objects.append(object)
                return objects;    

if __name__ == "__main__":
    main()
